src/Database/LSMTree/Internal.hs

```diff
@@ -1157,6 +1175,146 @@ mergeRuns resolve hfs hbio caching alloc runPaths mergeLevel runs = do
+      -- freed by the registry) the cursor state remains 'CursorClosed',
+      -- so the session (which might track the cursor) doesn't later find it
+      -- open and double frees resources.
+      modifyWithTempRegistry (pure CursorClosed) (void <$> swapMVar cursorState) $ \reg _ -> do
+        (cursorWriteBuffer, cursorRuns) <-
+          allocTableContent hfs hbio reg (tableContent thEnv)
+
+        mReaders <-
+          allocateMaybeTemp reg
+            (Readers.new hfs hbio (V.toList cursorRuns))
+            (Readers.close hfs hbio)
+
+        case mReaders of
+          Nothing -> do
+            freeTableContent hfs hbio reg cursorRuns
+            pure (CursorDrained, cursor)
+
+          Just cursorReaders -> do
+            cursorId <- incrUniqCounter (sessionUniqCounter cursorSessionEnv)
+            modifyMVar_ (sessionOpenCursors cursorSessionEnv) $
+              pure . Map.insert cursorId cursor
+            pure (CursorOpen CursorEnv {..}, cursor)
+  where
+    -- The table contents escape the read access, but we just added
+    -- references to each run, so it is safe.
+    allocTableContent hfs hbio reg contentVar = do
+        RW.withReadAccess contentVar $ \content -> do
+          let runs = cachedRuns (tableCache content)
+          V.forM_ runs $ \r -> do
+            allocateTemp reg
+              (Run.addReference hfs r)
+              (\_ -> Run.removeReference hfs hbio r)
+          pure (tableWriteBuffer content, runs)
+
+    -- We have to use 'freeTemp' to avoid double freeing on exception
+    freeTableContent hfs hbio reg runs = do
+        V.forM_ runs $ \r -> do
+          freeTemp reg (Run.removeReference hfs hbio r)
+
+{-# SPECIALISE closeCursor :: Cursor IO h -> IO () #-}
+-- | See 'Database.LSMTree.Normal.closeCursor'.
+closeCursor ::
+     m ~ IO -- TODO: replace by @io-classes@ constraints for IO simulation.
+  => Cursor m h
+  -> m ()
+closeCursor Cursor {..} = do
+    modifyMVar_ cursorState $ \case
+      CursorClosed -> return CursorClosed
+      CursorDrained -> return CursorClosed
+      CursorOpen CursorEnv {..} -> do
+        let hfs = sessionHasFS cursorSessionEnv
+        let hbio = sessionHasBlockIO cursorSessionEnv
+        -- TODO: what do we want to do if an exception is raised here?
+        -- We can't let 'modifyMVar_' put back 'CursorOpen' once we closed
+        -- parts of the cursor.
+        -- Do we want want close to ever leave 'CursorOpen' at all?
+        -- (same for closing table handles)
+
+        -- Since we have a write lock on the table state, we know that we are
+        -- the only thread currently closing the table. We can safely make the
+        -- session forget about this table.
+        modifyMVar_ (sessionOpenCursors cursorSessionEnv) $
+          pure . Map.delete cursorId
+        Readers.close hfs hbio cursorReaders
+        V.mapM_ (Run.removeReference hfs hbio) cursorRuns
+        return CursorClosed
```

Comment:
This now works the same as table handle close, but it doesn't seem right. Still thinking about it. Thoughts?

Reply:
Improved it, although probably still not perfect.